### PR TITLE
Fix potential long chain issues with Bernoulli treatment.

### LINKED ISSUE #24498: Bernoulli treatment may result in long call chains which exceed the number of ghosted layers.
## Bug Description
If multiple Bernoulli-jump faces are connected with only one cell they can result in a `number_of_faces+1` extra ghosting requirement depending on where the chain starts compared to the owned domain due to successive `GreenGaussGradient` computations. This results in errors if the number of ghosted cells cannot support the chain. This can get even worse for skewness-corrected cases.  

## Steps to Reproduce
Add more than 2 connected Bernoulli jump faces and start increasing the used ranks. Make sure you use fine enough mesh so the partitioning can get the a point where the first jump is on the other side of a ghosted cell.

## Impact
This can cause runtime termination if the partitioning of the mesh gives processor boundaries close to connected Bernoulli jumps.

### PR DESCRIPTION
**Pull request by grmnptr**
closes #24498

### CODE REVIEW

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md [hunk 1]
```diff
@@ -29,11 +29,28 @@ dimension explicit inlet boundary conditions related to velocity, one explicit
...
 from the explicit outlet pressure boundary condition, we believe it reasonable
 to do the same for computing the Bernoulli pressure jump at porosity
 discontinuities.
 
+!alert! note title=Behavior with parallel execution
+
+In certain cases multiple porosity jump faces may be connected by several cells in a
+chain. For example at corners or when porous medium zones are one-cell wide. In such
+scenarios, the two-term expansion in the downstream cell for the determination of the
+face pressure can lead to a very long chain of calls, which can potentially lead to
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:42]: let s not talk about chains of calls.
Maybe require an extended stencil in the direction of the porosity jumps?

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md [hunk 2]
```diff
@@ -29,11 +29,28 @@ dimension explicit inlet boundary conditions related to velocity, one explicit
...
+attempts of attempts of accessing cell values which are neither owned
+nor ghosted on the given rank. For this reason, the default value of
+[!param](/Variables/BernoulliPressureVariable/allow_two_term_expansion_on_bernoulli_faces)
+is `false`. If the user wants to enable two-term expansion for the pressure
+computation on the porosity ump faces, special attention should be paid to
+moving the porosity jump faces sufficiently far from each other or adding
+additional layers of ghosted elements (which can potentially increase communication costs).
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:49]: add the parameter ?

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md [hunk 3]
```diff
@@ -29,11 +29,28 @@ dimension explicit inlet boundary conditions related to velocity, one explicit
...
+In certain cases multiple porosity jump faces may be connected by several cells in a
+chain. For example at corners or when porous medium zones are one-cell wide. In such
+scenarios, the two-term expansion in the downstream cell for the determination of the
+face pressure can lead to a very long chain of calls, which can potentially lead to
+attempts of attempts of accessing cell values which are neither owned
+nor ghosted on the given rank. For this reason, the default value of
+[!param](/Variables/BernoulliPressureVariable/allow_two_term_expansion_on_bernoulli_faces)
+is `false`. If the user wants to enable two-term expansion for the pressure
+computation on the porosity ump faces, special attention should be paid to
+moving the porosity jump faces sufficiently far from each other or adding
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:48]: This isnt prescriptive enough. Do we need 2 cells in between each jump ?

File: modules/navier_stokes/include/base/NSFVBase.h [hunk 4]
```diff
@@ -796,6 +798,13 @@ NSFVBase<BaseType>::validParams()
       true,
       "If a two-term Taylor expansion is needed for the determination of the boundary values"
       "of the pressure.");
+  params.addParam<bool>(
+      "pressure_allow_expansion_on_bernoulli_faces",
+      false,
+      "Switch to enable the two-term extrapolation on porosity jump faces. "
+      "WARNING: This might lead to crushes in parallel runs if porosity jump faces are connected "
+      "with one cell (usually corners) due to the insufficient number of ghosted "
+      "layers.");
```
Comment (GiudGiud) [inline modules/navier_stokes/include/base/NSFVBase.h:807]: reword this in a positive way

File: modules/navier_stokes/test/tests/finite_volume/ins/action/errors/tests [hunk 5]
```diff
@@ -134,6 +134,20 @@
...
+    type = 'RunException'
+    input = 2d-rc-error-action.i
+    requirement = 'The system shall throw an error if the user tries to adjust the ghosting layers without using a Bernoulli porosity jump treatment.'
+    cli_args = "Modules/NavierStokesFV/pressure_allow_expansion_on_bernoulli_faces=true"
+    expect_err = "This parameter should not be given by the user with the corresponding porosity_interface_pressure_treatment setting!"
+  []
+  [no-bernoulli-two-term-error]
+    type = 'RunException'
+    input = 2d-rc-error-action.i
+    requirement = 'The system shall throw an error if two-term extrapolation is elected for prosity jump faces without enabling the Bernoulli treatment.'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/action/errors/tests:147]: the requirements are flipped between these two tests

File: modules/navier_stokes/test/tests/finite_volume/ins/action/errors/tests [hunk 6]
```diff
@@ -134,6 +134,20 @@
...
+    cli_args = "Modules/NavierStokesFV/ghost_layers=4"
+    expect_err = "This parameter should not be given by the user with the corresponding porosity_interface_pressure_treatment setting!"
+  []
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/action/errors/tests:150]: I d remove this test and this error. Users may want to set that for other reasons. I think being able to set ghost_layers is a net win

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md (hunk 2 above)
Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:49]: What parameter? The ghosting layer parameter from the kernels? I don't think we allow the user to tune those. If we are talking about the parameter to enable two term expansion, it is in the sentence before this.

Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:49]: Yeah the ghosting.
We may as well allow the user to set it imo, and have it in Advanced

Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:49]: I checked and we can actually set it, so we should be fine.

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md (hunk 3 above)
Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:48]: Well, unless we need skewness correction. Then we need more 3, added this detail.

File: modules/navier_stokes/include/base/NSFVBase.h [hunk 7]
```diff
@@ -796,6 +798,15 @@ NSFVBase<BaseType>::validParams()
       true,
       "If a two-term Taylor expansion is needed for the determination of the boundary values"
       "of the pressure.");
+  params.addParam<bool>(
+      "pressure_allow_expansion_on_bernoulli_faces",
+      false,
+      "Switch to enable the two-term extrapolation on porosity jump faces. "
+      "WARNING: Depending on the mesh, enabling this treatment parameter may lead to "
+      "termination in parallel runs due to insufficient ghosting between "
+      "processors. An example can be the presence of multiple porosity jumps connected with ony "
```
Comment (GiudGiud) [inline modules/navier_stokes/include/base/NSFVBase.h:807]:
```suggestion
      "processors. An example can be the presence of multiple porosity jumps connected with only "
```

File: modules/navier_stokes/include/base/NSFVBase.h [hunk 8]
```diff
@@ -841,12 +852,23 @@ NSFVBase<BaseType>::validParams()
...
+  /**
+   * Parameters controlling the ghosting/parallel execution
+   */
+  params.addRangeCheckedParam<unsigned short>(
+      "ghost_layers",
+      2,
+      "ghost_layers > 0",
+      "The number of geometric/algebraic/coupling layers to ghost.");
+
+  params.addParamNamesToGroup("ghost_layers", "Parallel Execution");
```
Comment (GiudGiud) [inline modules/navier_stokes/include/base/NSFVBase.h:870]:
```suggestion
  params.addParamNamesToGroup("ghost_layers", "Parallel Execution Tuning");
```

File: modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md [hunk 9]
```diff
@@ -42,7 +42,23 @@ commonly used in the action syntax is presented below:
 
   !listing include/base/NS.h line=std::string T_fluid
 
-For the default names of other variables used in this action, visit [this site](include/base/NS.h).  
+For the default names of other variables used in this action, visit [this site](include/base/NS.h).
+
+
+## Bernoulli pressure jump treatment
+
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md:49]: The problem here is that this summarizes something bigger with only its parallel issue.

I would rather you leave the section title but just leave a link to the right page:

```
## Bernoulli pressure jump treatment

Please see [the Bernouilli pressure variable documentation]([BernoulliPressureVariable.md](https://github.com/idaholab/moose/pull/24507/files#diff-e318d2c2747655bc9fde1436532f6abd19c131bf882fc39bde3c2edac2df172c)) for more information.

File: modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md [hunk 10]
```diff
@@ -42,7 +42,23 @@ commonly used in the action syntax is presented below:
...
+In certain cases multiple porosity jump faces may be connected by several cells in a
+chain. For example at corners or when porous medium zones are one-cell wide. In such
+scenarios, the two-term expansion in the downstream cell for the determination of the
+face pressure can lead to a very long chain of calls, which can potentially lead to
+attempts of attempts of accessing cell values which are neither owned
+nor ghosted on the given rank. For this reason, the default value of
+[!param](/Modules/NavierStokesFV/pressure_allow_expansion_on_bernoulli_faces)
+is `false`. If the user wants to enable two-term expansion for the pressure
+computation on the porosity ump faces, special attention should be paid to
```
Comment (lindsayad) [inline modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md:58]:
```suggestion
computation on the porosity jump faces, special attention should be paid to
```

File: modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md [hunk 11]
```diff
@@ -42,7 +42,23 @@ commonly used in the action syntax is presented below:
...
+moving the porosity jump faces sufficiently far from each other or adding
+additional layers of ghosted elements (which can potentially increase communication costs)
+using [!param](/Modules/NavierStokesFV/ghost_layers) parameter.
```
Comment (lindsayad) [inline modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md:61]:
```suggestion
additional layers of ghosted elements (which can potentially increase local computational and memory costs)
using [!param](/Modules/NavierStokesFV/ghost_layers) parameter.
```
Seems like we can just have this documentation in one place with requisite link from the other place

**Review by lindsayad — DISMISSED** [review]

File: modules/navier_stokes/src/variables/BernoulliPressureVariable.C [hunk 12]
```diff
@@ -128,24 +142,47 @@ BernoulliPressureVariable::getDirichletBoundaryFaceValue(const FaceInfo & fi,
...
       fi_elem_is_upwind ? bernoulli_vel_chunk_elem : bernoulli_vel_chunk_neighbor;
   const auto & downwind_bernoulli_vel_chunk =
       fi_elem_is_upwind ? bernoulli_vel_chunk_neighbor : bernoulli_vel_chunk_elem;
-  const auto p_downwind = (*this)(downwind_face, time);
+  ADReal p_downwind;
+  if (_allow_two_term_expansion_on_bernoulli_faces)
+    p_downwind = (*this)(downwind_face, time);
+  else
+    p_downwind = fi_elem_is_upwind ? (*this)(makeElemArg(fi.neighborPtr()), time)
+                                   : (*this)(makeElemArg(fi.elemPtr()), time);
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/src/variables/BernoulliPressureVariable.C:185]: I dont think we know yet. Historically we have been troubled in cases with frictions when the face extrapolation was turned off (big oscillation at inlet that Joe had iirc)

File: modules/navier_stokes/src/variables/BernoulliPressureVariable.C [hunk 13]
```diff
@@ -128,24 +142,47 @@ BernoulliPressureVariable::getDirichletBoundaryFaceValue(const FaceInfo & fi,
...
+  ADReal rho_neighbor;
+  if (_allow_two_term_expansion_on_bernoulli_faces)
+  {
+    rho_elem = (*_rho)(downwind_face, time);
+    rho_neighbor = rho_elem;
+  }
+  else
+  {
+    rho_elem = (*_rho)(makeElemArg(fi.elemPtr()), time);
+    rho_neighbor = (*_rho)(makeElemArg(fi.neighborPtr()), time);
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply modules/navier_stokes/src/variables/BernoulliPressureVariable.C:165]: Or at least I think we should always have the upstream rho determined from the cell center value of the upstream cell. It doesn't seem consistent to, in one case, use the downstream value for the upstream value, and then in another case use the upstream value

File: modules/navier_stokes/src/variables/BernoulliPressureVariable.C (hunk 12 above)
[the remark this answers is not in the document]
Comment (lindsayad) [reply modules/navier_stokes/src/variables/BernoulliPressureVariable.C:185]: Fair enough

File: modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md [hunk 14]
```diff
@@ -42,7 +42,23 @@ commonly used in the action syntax is presented below:
...
-For the default names of other variables used in this action, visit [this site](include/base/NS.h).  
+For the default names of other variables used in this action, visit [this site](include/base/NS.h).
+
+
+## Bernoulli pressure jump treatment
+
+In certain cases multiple porosity jump faces may be connected by several cells in a
+chain. For example at corners or when porous medium zones are one-cell wide. In such
+scenarios, the two-term expansion in the downstream cell for the determination of the
+face pressure can lead to a very long chain of calls, which can potentially lead to
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md:53]: This has been removed and now we are linking to the description in the BernoulliVariable.

File: modules/navier_stokes/src/variables/BernoulliPressureVariable.C (hunk 13 above)
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/navier_stokes/src/variables/BernoulliPressureVariable.C:165]: Addressed!

File: modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md (hunk 9 above)
Comment (grmnptr) [reply modules/navier_stokes/doc/content/syntax/Modules/NavierStokesFV/index.md:49]: Great idea!

**Review by GiudGiud — DISMISSED** [review]

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md [hunk 15]
```diff
@@ -36,17 +36,18 @@ discontinuities.
 
 !alert! note title=Behavior with parallel execution
 
-In certain cases multiple porosity jump faces may be connected by several cells in a
+In certain cases multiple porosity jump faces may be connected by single cells in a
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md]: You could just cut the word "single"

File: modules/navier_stokes/include/base/NSFVBase.h [hunk 16]
```diff
@@ -802,9 +802,11 @@ NSFVBase<BaseType>::validParams()
       "pressure_allow_expansion_on_bernoulli_faces",
       false,
       "Switch to enable the two-term extrapolation on porosity jump faces. "
-      "WARNING: This might lead to crushes in parallel runs if porosity jump faces are connected "
-      "with one cell (usually corners) due to the insufficient number of ghosted "
-      "layers.");
+      "WARNING: Depending on the mesh, enabling this treatment parameter may lead to "
```
Comment (lindsayad) [inline modules/navier_stokes/include/base/NSFVBase.h:805]: I would remove the word "treatment"

File: modules/navier_stokes/src/variables/BernoulliPressureVariable.C [hunk 17]
```diff
@@ -143,27 +143,14 @@ BernoulliPressureVariable::getDirichletBoundaryFaceValue(const FaceInfo & fi,
...
-    rho_elem = (*_rho)(downwind_face, time);
-    rho_neighbor = rho_elem;
-  }
-  else
-  {
-    rho_elem = (*_rho)(makeElemArg(fi.elemPtr()), time);
-    rho_neighbor = (*_rho)(makeElemArg(fi.neighborPtr()), time);
-  }
+  ADReal rho_elem = (*_rho)(makeElemArg(fi.elemPtr()), time);
+  ADReal rho_neighbor = (*_rho)(makeElemArg(fi.neighborPtr()), time);
```
Comment (lindsayad) [inline modules/navier_stokes/src/variables/BernoulliPressureVariable.C:153]: Nice. I like this change

File: modules/navier_stokes/include/base/NSFVBase.h (hunk 16 above)
Comment (grmnptr) [reply modules/navier_stokes/include/base/NSFVBase.h:805]: Done!

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md]: Removed!

File: modules/navier_stokes/include/base/NSFVBase.h [hunk 18]
```diff
@@ -802,9 +802,11 @@ NSFVBase<BaseType>::validParams()
...
+      "termination in parallel runs due to insufficient ghosting between "
+      "processors. An example can be the presence of multiple porosity jumps connected with only "
+      "one cell while using the Bernoulli pressure treatment. In such cases adjust the "
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/navier_stokes/include/base/NSFVBase.h:808]: Modified!

File: modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md [hunk 19]
```diff
@@ -36,17 +36,18 @@ discontinuities.
...
 [!param](/Variables/BernoulliPressureVariable/allow_two_term_expansion_on_bernoulli_faces)
 is `false`. If the user wants to enable two-term expansion for the pressure
-computation on the porosity ump faces, special attention should be paid to
-moving the porosity jump faces sufficiently far from each other or adding
-additional layers of ghosted elements (which can potentially increase communication costs).
+computation on the porosity jump faces, special attention should be paid to
+moving the porosity jump faces sufficiently far from each other (at least two layers
+if skewness correction is disabled and three if it is enabled) or adding
+additional layers of ghosted elements (which can potentially increase local computational and memory costs)
+using the following FVKernel parameter [!param](/FVKernels/FVDiffusion/ghost_layers).
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:50]: What does this mean? I cannot link to the parameters of an object which is not registered. Would you prefer to mention the parameter without linking to the description?

[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/navier_stokes/doc/content/source/variables/BernoulliPressureVariable.md:50]: Oh okay, I got it! Will add it, thanks!

**Review by lindsayad — APPROVED** [review]

**Pull request merged**